CMakeLists.txt

```diff
@@ -6,138 +6,555 @@
 
-cmake_minimum_required(VERSION 3.9.0)
+# FIXME (?) : downgraded requirement here to match the rest of LLVM.  There 
+# were no apparent issues but perhaps future PRs will change this? 
+cmake_minimum_required(VERSION 3.4.3) 
 
-option(LINK_WITH_FIR "Link driver with FIR and LLVM" ON)
+# RPATH settings on macOS do not affect INSTALL_NAME.
+if (POLICY CMP0068) 
+  cmake_policy(SET CMP0068 NEW)
+  set(CMAKE_BUILD_WITH_INSTALL_NAME_DIR ON)
+endif()
 
-# Pass -DGCC=... to cmake to use a specific gcc installation.
-if( GCC )
-  set(CMAKE_CXX_COMPILER "${GCC}/bin/g++")
-  set(CMAKE_CC_COMPILER "${GCC}/bin/gcc")
-  set(CMAKE_BUILD_RPATH "${GCC}/lib64")
-  set(CMAKE_INSTALL_RPATH "${GCC}/lib64")
-endif()
-if(BUILD_WITH_CLANG)
-  file(TO_CMAKE_PATH "${BUILD_WITH_CLANG}" CLANG_PATH)
-  set(CMAKE_CXX_COMPILER "${CLANG_PATH}/bin/clang++")
-  set(CMAKE_CC_COMPILER "${CLANG_PATH}/bin/clang")
-  if(GCC)
-    set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} --gcc-toolchain=${GCC}")
-  endif()
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wno-unused-command-line-argument")
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wstring-conversion")
-  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -Wcovered-switch-default")
+# From Clang: Include file check macros honor CMAKE_REQUIRED_LIBRARIES.
+if(POLICY CMP0075)   
+  cmake_policy(SET CMP0075 NEW)
 endif()
 
-# Set RPATH in every executable, overriding the default setting.
-# If you set this first variable back to true (the default),
-# also set the second one.
-set(CMAKE_SKIP_BUILD_RPATH false)
-set(CMAKE_BUILD_WITH_INSTALL_RPATH false)
+# option() honors normal variables.
+if (POLICY CMP0077) 
+  cmake_policy(SET CMP0077 NEW)
+endif()
 
-set(CMAKE_INSTALL_RPATH "${CMAKE_INSTALL_PREFIX}/lib" "${CMAKE_INSTALL_RPATH}")
+# Force all of Flang into C++17 mode... 
+# FIXME (?) : I've had no issues with building across numerous platforms with 
+# the base components of LLVM build in C++14 and Flang with C++17 but 
+# this might bite someone as we start tighter integration (e.g. w/ MLIR). 
+set(CMAKE_CXX_STANDARD 17)
 
-# Reminder: Setting CMAKE_CXX_COMPILER must be done before calling project()
+# Check to see if are part of an LLVM build, if not we need 
+# to configure as a standalone project (use LLVM as an external 
+# library).
+if (CMAKE_SOURCE_DIR STREQUAL CMAKE_CURRENT_SOURCE_DIR)
+  project(Flang)
 
-project(f18 CXX)
+  set(CONFIG_OUTPUT)
+  if (LLVM_CONFIG)
 
-if( NOT CMAKE_BUILD_TYPE )
-  set( CMAKE_BUILD_TYPE Debug )
-endif()
+    set(LLVM_CONFIG_FOUND 1)
+    message(STATUS "Found llvm_config as ${LLVM_CONFIG}.")
+    message(DEPRECATION "Using llvm-config to detect the LLVM instalation is \
+            deprecated.  The installed cmake files should be used instead. \
+            CMake should be able to detect your LLVM install automatically, \
+            but you can also use LLVM_DIR to specify the path to the directory \
+            containing LLVMConfig.cmake.")
```

Comment:
If using llvm-config is deprecated, why add support for it? The cmake file would be simpler if that support were eliminated.

Reply:
Yes, in big picture I agree. From what I could see at the time is that this style has not yet been fully deprecated in Clang so I followed (what appears to be) the community's current practice here.

Reply:
No one is currently using LLVM_CONFIG to build f18 and it has been deprecated in clang for more than a year. Can we wait until someone requests it before adding support for it? Otherwise this code is all waste.